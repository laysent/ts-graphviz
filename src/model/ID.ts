import { DotBase } from '../abstract';
import { IID } from '../types';
import { escape, quote } from '../utils/dot-rendering';

/**
 * An ID is just a value.
 * Designed as an immutable object.
 *
 * @description
 * An ID is one of the following:
 * - Any string of alphabetic ([a-zA-Z\200-\377]) characters, underscores ('_') or digits ([0-9]), not beginning with a digit;
 * - a numeral [-]?(.[0-9]+ | [0-9]+(.[0-9]*)? );
 * - any double-quoted string ("...") possibly containing escaped quotes (\")1;
 * - an HTML string (<...>).
 *
 * @category Internal
 */
export class ID extends DotBase implements IID {
  public readonly value: string;

  /** @hidden */

  private isHTMLLike: boolean;
  /** @hidden */
  private isNotString: boolean;
  /** @hidden */
  private isQuoteRequired: boolean;

  constructor(value: string | number | boolean) {
    super();
    this.isNotString = typeof value !== 'string';
    let stringValue: string = typeof value === 'string' ? value : value.toString();
    if (this.isNotString) {
      this.isHTMLLike = false;
      this.isQuoteRequired = false;
    } else {
      const trimmed = stringValue.trim();
      this.isHTMLLike = /^<.+>$/ms.test(trimmed);
      if (this.isHTMLLike) {
        stringValue = trimmed;
        this.isQuoteRequired = false;
      } else {
        this.isQuoteRequired = true;
      }
    }
    this.value = stringValue;
  }

  /**
   * Determines whether quotes are required or HTML-Like labels and returns an appropriate string.
   */
  public toDot(): string {
    if (this.isNotString || this.isHTMLLike) {
      return this.value;
    }
    let value = this.value;
    if (this.isQuoteRequired) {
      value = quote(escape(value));
    }
    return value;
  }
}
