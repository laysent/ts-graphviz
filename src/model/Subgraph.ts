import { attribute } from '../attribute';
import { ClusterType, IContext, ISubgraph } from '../types';
import { Cluster } from './Cluster';

/**
 * Subgraph object.
 * @category Primary
 */
export class Subgraph extends Cluster<attribute.Subgraph | attribute.ClusterSubgraph> implements ISubgraph {
  /** Indicates the type of cluster. */
  public type = ClusterType.subgraph;
  public attributes = {
    graph: this.context.createAttributes<attribute.ClusterSubgraph>(),
    edge: this.context.createAttributes<attribute.Edge>(),
    node: this.context.createAttributes<attribute.Node>(),
  };
  constructor(public readonly context: IContext) {
    super();
  }
  /** Determines whether the Subgraph is a SubgraphCluster. */
  public isSubgraphCluster(): boolean {
    if (typeof this.id === 'string') {
      return this.id.startsWith('cluster');
    }
    return false;
  }
}
