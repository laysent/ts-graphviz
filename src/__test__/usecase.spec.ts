import 'jest-graphviz';
import { Digraph, Graph } from '../model';
import { digraph, graph, strict } from '../usecase';

describe('function digraph', () => {
  it('should return Digraph object, when execute digraph()', () => {
    const g = digraph();
    expect(g).toBeInstanceOf(Digraph);
  });

  test('callback style', () => {
    const G = digraph('G', g => {
      const a = g.node('aa');
      const b = g.node('bb');
      const c = g.node('cc');
      g.edge([a, b, c], e => {
        e.attributes.set('color', 'red');
      });
      g.subgraph('A', A => {
        const Aa = A.node('Aaa', n => {
          n.attributes.set('color', 'pink');
        });
        const Ab = A.node('Abb', n => {
          n.attributes.set('color', 'violet');
        });
        const Ac = A.node('Acc');
        A.edge([Aa.port('a'), Ab, Ac, 'E'], e => {
          e.attributes.set('color', 'red');
        });
      });
    });
    const dot = G.toDot();
    expect(dot).toBeValidDotAndMatchSnapshot();
  });

  test('comment', () => {
    const G = digraph('G', g => {
      g.comment = 'This is directed graph.';
      const a = g.node('aa');
      a.comment = 'This is node a.';
      const b = g.node('bb');
      b.comment = 'This is node b.';
      const c = g.node('cc');
      c.comment = 'This is node c.';
      g.edge([a, b, c], e => {
        e.comment = 'This is edge.\nIt connects a, b and c.';

        e.attributes.set('color', 'red');
        e.attributes.comment = 'Edge line will draw with red.';
      });
      g.subgraph('A', A => {
        A.comment = 'It is subgraph A.\nIt is not cluster';
        A.node('Aaa', n => {
          n.comment = 'This is node Aaa in subgraph A.';
          n.attributes.comment = 'It will be filled by pink.';
          n.attributes.set('color', 'pink');
        });
      });
    });
    const dot = G.toDot();
    expect(dot).toBeValidDotAndMatchSnapshot();
  });
});

describe('function graph', () => {
  it('should return Graph object, when execute graph()', () => {
    const g = graph();
    expect(g).toBeInstanceOf(Graph);
  });

  test('callback style', () => {
    const G = graph('G', g => {
      const a = g.node('aa');
      const b = g.node('bb');
      const c = g.node('cc');
      g.edge([a, b, c], e => {
        e.attributes.set('color', 'red');
      });
      g.subgraph('A', A => {
        const Aa = A.node('Aaa', n => {
          n.attributes.set('color', 'pink');
        });
        const Ab = A.node('Abb', n => {
          n.attributes.set('color', 'violet');
        });
        const Ac = A.node('Acc');
        A.edge([Aa, Ab, Ac, A.node('hoge').port('fuga')], e => {
          e.attributes.set('color', 'red');
        });

        A.edge([Aa.port({ port: 'a', compass: 'w' }), Ab.port({ compass: 'w' }), 'Aaa:e', 'Acc:r:e'], e => {
          e.attributes.set('color', 'red');
        });
      });
    });
    const dot = G.toDot();
    expect(dot).toBeValidDotAndMatchSnapshot();
  });

  test('escape characters', () => {
    const G = graph('G\n"', g => {
      const a = g.node('a\na');
      a.attributes.set('label', '1\n2\n"\n3');
      const b = g.node('b"b');
      const c = g.node('c\nc"');
      g.edge([a, b, c], e => {
        e.attributes.set('color', 'red');
      });

      g.subgraph('graph.name', s => {
        const innerA = s.node('node.name');
        innerA.attributes.set('label', 'node');
        const innerB = s.node('another.name');
        innerB.attributes.set('label', 'words with space and "quote"');
        s.edge([innerA, innerB]);
      });
    });
    const dot = G.toDot();
    expect(dot).toBeValidDotAndMatchSnapshot();
  });
});

describe('strict mode', () => {
  it('should return Digraph object, when execute digraph()', () => {
    const g = strict.digraph();
    expect(g.strict).toBe(true);
    expect(g).toBeInstanceOf(Digraph);
  });

  it('should return Graph object, when execute graph()', () => {
    const g = strict.graph();
    expect(g.strict).toBe(true);
    expect(g).toBeInstanceOf(Graph);
  });
});
