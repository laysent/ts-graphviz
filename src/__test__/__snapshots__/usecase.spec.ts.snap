// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`function digraph callback style 1`] = `
"digraph \\"G\\" {
  \\"aa\\";
  \\"bb\\";
  \\"cc\\";
  subgraph \\"A\\" {
    \\"Aaa\\" [
      color = \\"pink\\",
    ];
    \\"Abb\\" [
      color = \\"violet\\",
    ];
    \\"Acc\\";
    \\"E\\";
    \\"Aaa\\":\\"a\\" -> \\"Abb\\" -> \\"Acc\\" -> \\"E\\" [
      color = \\"red\\",
    ];
  }
  \\"aa\\" -> \\"bb\\" -> \\"cc\\" [
    color = \\"red\\",
  ];
}"
`;

exports[`function digraph comment 1`] = `
"// This is directed graph.
digraph \\"G\\" {
  // This is node a.
  \\"aa\\";
  // This is node b.
  \\"bb\\";
  // This is node c.
  \\"cc\\";
  // It is subgraph A.
  // It is not cluster
  subgraph \\"A\\" {
    // This is node Aaa in subgraph A.
    \\"Aaa\\" [
      // It will be filled by pink.
      color = \\"pink\\",
    ];
  }
  // This is edge.
  // It connects a, b and c.
  \\"aa\\" -> \\"bb\\" -> \\"cc\\" [
    // Edge line will draw with red.
    color = \\"red\\",
  ];
}"
`;

exports[`function graph callback style 1`] = `
"graph \\"G\\" {
  \\"aa\\";
  \\"bb\\";
  \\"cc\\";
  subgraph \\"A\\" {
    \\"Aaa\\" [
      color = \\"pink\\",
    ];
    \\"Abb\\" [
      color = \\"violet\\",
    ];
    \\"Acc\\";
    \\"hoge\\";
    \\"Aaa\\" -- \\"Abb\\" -- \\"Acc\\" -- \\"hoge\\":\\"fuga\\" [
      color = \\"red\\",
    ];
    \\"Aaa\\":\\"a\\":w -- \\"Abb\\":w -- \\"Aaa\\":\\"e\\" -- \\"Acc\\":\\"r\\":e [
      color = \\"red\\",
    ];
  }
  \\"aa\\" -- \\"bb\\" -- \\"cc\\" [
    color = \\"red\\",
  ];
}"
`;

exports[`function graph escape characters 1`] = `
"graph \\"G\\\\n\\\\\\"\\" {
  \\"a\\\\na\\" [
    label = \\"1\\\\n2\\\\n\\\\\\"\\\\n3\\",
  ];
  \\"b\\\\\\"b\\";
  \\"c\\\\nc\\\\\\"\\";
  subgraph \\"graph.name\\" {
    \\"node.name\\" [
      label = \\"node\\",
    ];
    \\"another.name\\" [
      label = \\"words with space and \\\\\\"quote\\\\\\"\\",
    ];
    \\"node.name\\" -- \\"another.name\\";
  }
  \\"a\\\\na\\" -- \\"b\\\\\\"b\\" -- \\"c\\\\nc\\\\\\"\\" [
    color = \\"red\\",
  ];
}"
`;
